pep-0675.rst

```diff
@@ -0,0 +1,954 @@
+
+    literal_str: Literal[str] = "hello"
+    x: Container[Literal[str]] = Container(literal_str)  # OK
+
+    s: str
+    x_error: Container[Literal[str]] = Container(s)  # Not OK
+
+Standard containers like ``List`` work as expected:
+
+::
+
+    xs: List[Literal[str]] = ["foo", "bar", "baz"]
+
+Interactions with Overloads
+---------------------------
+
+Literal strings and overloads do not need to interact in a special
+way: the existing rules work fine. ``Literal[str]`` can be used as a
+fallback overload where a specific ``Literal["foo"]`` type does not
+match:
+
+::
+
+    @overload
+    def foo(x: Literal["foo"]) -> int: ...
+    @overload
+    def foo(x: Literal[str]) -> bool: ...
+    @overload
+    def foo(x: str) -> str: ...
+
+    x1: int = foo("foo")  # First overload.
+    x2: bool = foo("bar")  # Second overload.
+    s: str
+    x3: str = foo(s)  # Third overload.
+
+Backwards Compatibility
+-----------------------
+
+As PEP 586 `mentions
+<https://www.python.org/dev/peps/pep-0586/#backwards-compatibility>`_,
+type checkers "should feel free to experiment with more sophisticated
+inference techniques". So, if the type checker infers a literal string
+type for an unannotated variable that is initialized with a literal
+string, the following example should be OK:
+
+::
+
+    x = "hello"
+    expect_literal_str(x)
+    # OK, because x is inferred to have type ``Literal["hello"]``.
+
+This enables precise type checking of idiomatic SQL query code without
+annotating the code at all (as seen in the `Motivation`_ section
+example).
+
+However, like PEP 586, this PEP does not mandate the above inference
+strategy. In case the type checker doesn't infer ``x`` to have type
+``Literal["hello"]``, users can aid the type checker by explicitly
+annotating it as ``x: Literal[str]``:
+
+::
+
+    x: Literal[str] = "hello"
+    expect_literal_str(x)
+
+Runtime behavior
+================
+
+This PEP does not change the runtime behavior of ``Literal``.
+
+Backwards compatibility
+=======================
+
+Backwards compatibility: ``Literal[str]`` is acceptable at runtime, so
+this doesn't require any changes to the Python runtime itself. PEP 586
+already backports ``Literal``, so this PEP does not need to change it.
+
+
+Rejected Alternatives
+=====================
+
+Why not use tool X?
+-------------------
+
+Focusing solely on the example of preventing SQL injection, tooling to
+catch this kind of issue seems to come in three flavors: AST based,
+function level analysis, and taint flow analysis.
+
+**AST based tools include Bandit and pyanalyze**: `Bandit
+<https://github.com/PyCQA/bandit/blob/aac3f16f45648a7756727286ba8f8f0cf5e7d408/bandit/plugins/django_sql_injection.py#L102>`_
+has a plugin to warn when SQL queries are not literal
+strings. `pyanalyze
+<https://github.com/quora/pyanalyze/blob/afcb58cd3e967e4e3fea9e57bb18b6b1d9d42ed7/README.md#extending-pyanalyze>`_
+has a similar extension. The problem is that many perfectly safe SQL
```

Comment:
Pyanalyze in fact handles the example in the motivation:
```
Revealed type is 'Literal['xy', 'x']' (code: inference_failure)
In striadd.py at line 5
   2:     x = "x"
   3:     if cond:
   4:         x += "y"
   5:     reveal_type(x)
                      ^
```
It's true though that it doesn't handle the full range of what `Literal[str]` would allow (e.g., having a function of type `(Literal[str]) -> Literal[str]`). In that regard it looks more like semgrep than bandit; I haven't used either of those other tools though.

Reply:
Thanks for confirming! We weren't sure about `pyanalyze`. Given your explanation yesterday, I realize that `pyanalyze` is more runtime-based than AST-based. Moved it to the Semgrep bullet.